modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/slot/SlotManager.java

```diff
@@ -57,27 +61,59 @@
     private IMap<String, Long> queueToLastAssignedIDMap;
 
     /**
+     * To keep track of last published ID against node
+     */
+    private IMap<String, Long> nodeToLastPublishedIDMap;
+
+    /**
      * To keep track of assigned slots up to now. Key of the map contains nodeID. Value is
      * HashmapStringListWrapper object. HashmapStringListWrapper is a wrapper class for
      * HashMap<String,List<Slot>>. Key in that hash map is queue name. value is List of Slot
      * objects.
      */
-    private IMap<String, HashmapStringListWrapper> slotAssignmentMap;
+    private IMap<String, HashmapStringTreeSetWrapper> slotAssignmentMap;
+
+    /**
+     * HazelCast map keeping overlapped slots. Overlapped slots are eligible to be assigned again.
+     * Key of the map contains nodeID. Value is
+     * HashMapStringListWrapper object. HashMapStringListWrapper is a wrapper class for
+     * HashMap<String,List<Slot>>. Key in that hash map is queue name. value is List of Slot
+     * objects.
+     */
+    private IMap<String, HashmapStringTreeSetWrapper> overLappedSlotMap;
+
+    /**
+     * Keeps slot deletion safe zones for each node, as they are informed by nodes via thrift
+     * communication. Used for safe zone calculation for cluster by Slot Manager
+     */
+    private Map<String, Long> nodeInformedSlotDeletionSafeZones;
+
+    //safe zone calculator
+    private SlotDeleteSafeZoneCalc slotDeleteSafeZoneCalc;
 
     private static Log log = LogFactory.getLog(SlotManager.class);
 
 
     private SlotManager() {
         if (AndesContext.getInstance().isClusteringEnabled()) {
+
             HazelcastAgent hazelcastAgent = HazelcastAgent.getInstance();
-            /**
-             * Initialize distributed maps used in this class
-             */
+
+            //Initialize distributed maps used in this class
             unAssignedSlotMap = hazelcastAgent.getUnAssignedSlotMap();
+            overLappedSlotMap = hazelcastAgent.getOverLappedSlotMap();
             slotIDMap = hazelcastAgent.getSlotIdMap();
             queueToLastAssignedIDMap = hazelcastAgent.getLastAssignedIDMap();
+            nodeToLastPublishedIDMap = hazelcastAgent.getLastPublishedIDMap();
             slotAssignmentMap = hazelcastAgent.getSlotAssignmentMap();
 
+            nodeInformedSlotDeletionSafeZones = new HashMap<String, Long>();
+
+            //start a thread to calculate slot delete safe zone
+            //TODO: use a common  thread pool for tasks like this?
+            slotDeleteSafeZoneCalc = new SlotDeleteSafeZoneCalc(SAFE_ZONE_EVALUATION_INTERVAL);
+            new Thread(slotDeleteSafeZoneCalc).start();
```

Comment:
As mentioned in the comment great if we can get it into a thread-pool and then name the pool as well. Would come in handy when debugging